# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: browser/components/BrowserGlue.jsm
# Commit: 626dd2a045af
# Full Hash: 626dd2a045af6cd510ef04e67d027909509547f1
# Author: Mike Conley <mconley@mozilla.com>
# Date: 2020-04-17 21:46:50
# Regressor Bug: 1614502
# File Overlap Count: 2
# Description:
#   Bug 1614502 - Add a mechanism that can generate an about:home document for a startup cache. r=Gijs,k88hudson
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D68094
# ==============================================================================

diff -r 7ec776d31a2c -r 626dd2a045af browser/components/BrowserGlue.jsm
--- a/browser/components/BrowserGlue.jsm	Fri Apr 17 15:31:11 2020 +0000
+++ b/browser/components/BrowserGlue.jsm	Fri Apr 17 15:31:39 2020 +0000
@@ -4793,6 +4793,9 @@
   // should load.
   SEND_STREAMS_MESSAGE: "AboutHomeStartupCache:InputStreams",
 
+  STATE_RESPONSE_MESSAGE: "AboutHomeStartupCache:State:Response",
+  STATE_REQUEST_MESSAGE: "AboutHomeStartupCache:State:Request",
+
   // A reference to the nsICacheEntry to read from and write to.
   _cacheEntry: null,
 
@@ -4835,6 +4838,7 @@
 
     Services.obs.addObserver(this, "ipc:content-created");
     Services.ppmm.addMessageListener(this.POPULATE_MESSAGE, this);
+    Services.ppmm.addMessageListener(this.STATE_REQUEST_MESSAGE, this);
 
     this.log = Log.repository.getLogger(this.LOG_NAME);
     this.log.manageLevelFromPref(this.LOG_LEVEL_PREF);
@@ -4864,6 +4868,7 @@
 
     Services.obs.removeObserver(this, "ipc:content-created");
     Services.ppmm.removeMessageListener(this.POPULATE_MESSAGE, this);
+    Services.ppmm.removeMessageListener(this.STATE_REQUEST_MESSAGE, this);
     this._pagePipe = null;
     this._scriptPipe = null;
     this._initted = false;
@@ -5108,9 +5113,18 @@
       return;
     }
 
-    if (message.name == this.POPULATE_MESSAGE) {
-      let { pageInputStream, scriptInputStream } = message.data;
-      this.populateCache(pageInputStream, scriptInputStream);
+    switch (message.name) {
+      case this.POPULATE_MESSAGE: {
+        let { pageInputStream, scriptInputStream } = message.data;
+        this.populateCache(pageInputStream, scriptInputStream);
+        break;
+      }
+      case this.STATE_REQUEST_MESSAGE: {
+        this.log.trace("Parent got request for Activity Stream state object.");
+        let state = AboutNewTab.activityStream.store.getState();
+        message.target.sendAsyncMessage(this.STATE_RESPONSE_MESSAGE, { state });
+        break;
+      }
     }
   },
 